proposals/gc/MVP.md

```diff
@@ -266,62 +293,112 @@ Tentatively, support a type of guaranteed unboxed scalars.
+  - `br_on_func $l : [anyref] -> [anyref]`
+    - iff `$l : [i31ref]`
+  - passes operand along with branch as a function
+
+* `ref.as_func` converts to a function reference
+  - `ref.as_func : [anyref] -> [funcref]`
+  - traps if reference is not a function
+  - equivalent to `(block $l (param anyref) (result funcref) (br_on_func $l) (unreachable))`
+
+* `ref.as_data` converts to a data reference
+  - `ref.as_data : [anyref] -> [dataref]`
+  - traps if reference is not compound data
+  - equivalent to `(block $l (param anyref) (result dataref) (br_on_data $l) (unreachable))`
 
 * `ref.as_i31` converts to an integer reference
   - `ref.as_i31 : [anyref] -> [i31ref]`
   - traps if reference is not an integer
-  - equivalent to `(rtt.canon i31) (ref.cast)`
+  - equivalent to `(block $l (param anyref) (result i31ref) (br_on_i31 $l) (unreachable))`
+
+Note: There are no instructions to check for `externref`, since that can consist of a diverse set of different object representations that would be costly to check for exhaustively.
+
+TODO: Should we add `br_on_null` for completeness?
 
 
 #### Runtime Types
 
-* `rtt.canon <heaptype>` returns the RTT of the specified type
-  - `rtt.canon t : [] -> [(rtt n t)]`
-  - `n = 0` iff `t = any`, and `n = 1` otherwise
+* `rtt.canon <typeidx>` returns the RTT of the specified type
+  - `rtt.canon $t : [] -> [(rtt 0 $t)]`
   - multiple invocations of this instruction yield the same observable RTTs
   - this is a *constant instruction*
-  - equivalent to `(rtt.sub 1 any t (rtt.canon any))`, except when `t` itself is `any`
 
-* `rtt.sub <n> <heaptype1> <heaptype2>` returns an RTT for `heaptype2` as a sub-RTT of a the parent RTT operand for `heaptype1`
-  - `rtt.sub n t1 t2 : [(rtt n t1)] -> [(rtt (n+1) t2)]`
+* `rtt.sub <n> <typeidx1> <typeidx2>` returns an RTT for `typeidx2` as a sub-RTT of a the parent RTT operand for `typeidx1`
+  - `rtt.sub n $t1 $t2 : [(rtt n $t1)] -> [(rtt (n+1) $t2)]`
     - iff `t2 <: t1`
   - multiple invocations of this instruction with the same operand yield the same observable RTTs
   - this is a *constant instruction*
 
-TODO: Add the ability to generate new (non-canonical) RTT values to implement casting in nominal type hierarchies.
+TODO: Add the ability to generate new (non-canonical) RTT values to implement casting in nominal type hierarchies?
 
 
 #### Casts
 
-* `ref.test <heaptype1> <heaptype2>` tests whether a reference value's [runtime type](#values) is a [runtime subtype](#runtime) of a given RTT
-  - `ref.test t1 t2 : [(ref null t1) (rtt n t2)] -> [i32]`
-     - iff `t2 <: t1`
+RTT-based casts can only be performed with respect to concrete types, and require a data or function reference as input, which are known to carry an RTT.
+
+* `ref.test <typeidx>` tests whether a reference value's [runtime type](#values) is a [runtime subtype](#runtime) of a given RTT
+  - `ref.test $t : [(ref null ht) (rtt n $t)] -> [i32]`
```

Comment:
I think this should be `(rtt n? $t)`, i.e. the depth `n` is optional.

Reply:
Yup, fixed here and below.